### ISSUE TITLE
Publisher MR complete but  bin/gobblin-mapreduce.sh still running.

### ISSUE THREAD

**Issue by krtiwari1981** [linkedin/gobblin#844]
Hello,

I am running on MR mode , i can see in the resource manager the job is complete and the data is copied to the S3 location but call to bin/gobblin-mapreduce.sh is running for very long time and keeps on logging the below message in gobblin-current.log. The job finishes but it takes very long time at this step.

In my case the below message is logged for 15 minutes. 

Can somebody point why this is happening ? 

2016-03-14 21:29:41 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:41 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:42 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:42 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:43 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:43 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:43 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:43 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:44 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:44 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:44 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:45 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:45 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:45 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:45 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:46 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:47 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:47 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:47 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:48 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:48 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:48 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:49 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:49 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:49 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-14 21:29:49 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-

**Comment by sahilTakiar** [comment]
Can you send your config file, as well as a full copy of the logs. I suspect this is related to some cleanup done by the `S3AFileSystem`.

**Comment by krtiwari1981** [comment]
Attached are config and log

[log.txt](https://github.com/linkedin/gobblin/files/172949/log.txt)
[config.txt](https://github.com/linkedin/gobblin/files/172948/config.txt)

**Comment by sahilTakiar** [comment]
A few comments:

* I notice you have `state.store.fs.uri` set twice, the first time it points to a s3 bucket, the second time it points to HDFS; I am not sure which one Gobblin will use
    * If the s3 bucket is used for the state-store this could explain the behavior you are seeing (see below)
* The logs you are seeing are most likely a result of invoking `S3AFileSystem.create()` - which creates a `S3AOutputStream`
    * A `S3AOutputStream` uses the `LocalDirAllocator` to create a directory on the local fs where it stages data before uploading it to S3
        * Check out http://grepcode.com/file/repo1.maven.org/maven2/org.apache.hadoop/hadoop-aws/2.6.0/org/apache/hadoop/fs/s3a/S3AOutputStream.java#S3AOutputStream and http://gobblin.readthedocs.org/en/latest/case-studies/Publishing-Data-to-S3/ for more info
    * This is why these logs keep on getting printed out
* Try removing `state.store.fs.uri=s3a://mybucket/` and see if the behavior changes
    * If the `state.store.fs.uri` uses `s3a` the data first gets written to HDFS, then to the local fs, and then uploaded to s3 which is why it takes so long, if you remove this things should change

**Comment by sahilTakiar** [comment]
I also noticed you wrote a few custom classes for this job, if there is any logic you think can be contributed back to Gobblin, we always welcome external contributions. Submitting a PR is pretty easy: https://help.github.com/articles/creating-a-pull-request/

**Comment by krtiwari1981** [comment]
Hi Sahil,

My required to store data in S3. 
You are right the state is getting stored in s3.
So i removed the state.store.fs.uri pointing to hdfs hostname and trying to run again to check how it is behaving.

Earlier it took 10 minutes to pull the data and 15 minutes to copy . 

My process seem to be very slow and is not meeting current SLA, I guess from your explanation it seems bulk of the time is spent copying the data to S3. 

Do you pointers to increase the performance?. 

Once i am done done with the project, surely i will contribute back.

**Comment by sahilTakiar** [comment]
Did changing the state-store filesystem make a significant difference? It should have decreases the amount of time taken by the job significantly

**Comment by krtiwari1981** [comment]
i have to run few more tests to check, it seems to help.

But in the log i still see ,

2016-03-15 01:12:16 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-
2016-03-15 01:12:16 UTC INFO  [ParallelRunner] org.apache.hadoop.fs.LocalDirAllocator$AllocatorPerContext$DirSelector  530 - Returning directory: /media/ephemeral0/hadoop-root/s3a/output-

how can i change the path to something else from /media/ephemeral0/hadoop-root/s3a/output-

 i have configured the config to both store data and state in s3.

The doc says 

Thus, when using the s3a file system with Gobblin it is recommended that one configures Gobblin to first write its staging data to the local filesystem, and then to publish the data to S3. 

Is this property that the above posts talks about 
**writer.staging.dir
**

or there is something 
**fs.s3a.staging.dir
**

Basically i want to change the location /media/ephemeral0/hadoop-root/s3a/output-

to something else i don't know where i should change.

I guess this /media/ephemeral0/hadoop-root/s3a/output- is slow disk so i want to change it.

**Comment by krtiwari1981** [comment]
i figured out the property is 

fs.s3a.buffer.dir

**Comment by sahilTakiar** [comment]
@krtiwari1981 I am closing this ticket, please re-open if there are any follow up items. I updated the `gobblin-docs` to include documentation of this property.
